pkg/httpclient/client_test.go

```diff
@@ -0,0 +1,144 @@
+	"strings"
+	"testing"
+
+	"github.com/dcos/dcos-cli/pkg/config"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestNew(t *testing.T) {
+	acsToken := "acsToken"
+	url := "https://dcos.io"
+
+	conf := config.New()
+	conf.SetURL("https://dcos.io")
+	conf.SetACSToken(acsToken)
+
+	client := New(conf)
+
+	tlsConfig := &tls.Config{
+		InsecureSkipVerify: conf.TLS().Insecure,
+		RootCAs:            conf.TLS().RootCAs,
+	}
+	transport := &http.Transport{
+		TLSClientConfig: tlsConfig,
+	}
+	backendClient := &http.Client{
+		Transport: transport,
+		Timeout:   conf.Timeout(),
+	}
+	expectedClient := &Client{
+		acsToken: acsToken,
+		baseURL:  url,
+		client:   backendClient,
+	}
+
+	require.Equal(t, expectedClient, client)
+}
+
+func TestGet(t *testing.T) {
+	body := `{"cluster":"DC/OS"}`
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != "GET" {
+			t.Errorf("Expected `GET` request, got `%s`", r.Method)
+		}
+		if r.URL.EscapedPath() != "/get" {
+			t.Errorf("expected request to `/get, got `%s`", r.URL.EscapedPath())
+		}
+
+		w.Write([]byte(body))
+	}))
+	defer ts.Close()
+
+	conf := config.New()
+	conf.SetURL(ts.URL)
+
+	client := New(conf)
+
+	resp, err := client.Get("/get")
+	require.NoError(t, err)
+
+	respBody, err := ioutil.ReadAll(resp.Body)
+	require.NoError(t, err)
+	require.Equal(t, body, string(respBody))
+}
+
+func TestPost(t *testing.T) {
+	type bodyStruct struct {
+		Cluster string
+	}
+	body := `{"cluster":"DC/OS"}`
+
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != "POST" {
+			t.Errorf("Expected `POST` request, got `%s`", r.Method)
+		}
+		if r.URL.EscapedPath() != "/post" {
+			t.Errorf("expected request to `/post, got `%s`", r.URL.EscapedPath())
+		}
+
+		body, err := ioutil.ReadAll(r.Body)
+		require.NoError(t, err)
+
+		assert.Equal(t, string(body), `{"cluster":"DC/OS"}`)
+	}))
+	defer ts.Close()
+
+	conf := config.New()
+	conf.SetURL(ts.URL)
+
+	client := New(conf)
+
+	resp, err := client.Post("/post", "application/json", strings.NewReader(body))
+	require.NoError(t, err)
+	require.Equal(t, resp.StatusCode, http.StatusOK)
+}
+
+func TestACSToken(t *testing.T) {
+	body := `{"cluster":"DC/OS"}`
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != "GET" {
+			t.Errorf("Expected `GET` request, got `%s`", r.Method)
+		}
+		if r.Header.Get("Authorization") != "token=secret" {
+			t.Errorf("expected Authorization header to be `secret`, got `%s`", r.Header.Get("Authorization"))
+		}
+		if r.URL.EscapedPath() != "/get" {
+			t.Errorf("expected request to `/get, got `%s`", r.URL.EscapedPath())
+		}
+
+		w.Write([]byte(body))
+	}))
+	defer ts.Close()
+
+	conf := config.New()
+	conf.SetURL(ts.URL)
+	conf.SetACSToken("secret")
+
+	client := New(conf)
+
+	_, err := client.Get("/get")
+	require.NoError(t, err)
```

Comment:
If the `client.Get()` method were not doing any HTTP request at all (eg. just `return nil, nil`), the test would still succeed.

We are making assertions in the test HTTP handler func without making sure that the handler function is actually called. This can be done by either returning a response with a well-known message and asserting that the client returns the expected reponse, or in a more generic way by having the handler setting a flag or something indicating it has been called.